scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala

```diff
@@ -1909,6 +1909,39 @@ class FormatOps(
     }
   }
 
+  object OptionalBraces {
+    def unapply(ft: FormatToken): Option[(FormatToken, Tree)] = ft.left match {
+      case _ if !dialect.allowSignificantIndentation => None
+      case _: T.Comment | _: T.LeftBrace => None
+      case _: T.Colon => // Optional braces in templates after `:`
+        ft.meta.leftOwner match {
+          case t: Template =>
+            val ok = t.tokens.headOption.contains(ft.left)
+            if (ok) Some((ft, t)) else None
+          case _ => None
+        }
+      case _: T.KwWith => // Optional braces in templates after `with`
+        ft.meta.leftOwner match {
+          case t: Template =>
+            val nextFt = nextNonComment(ft)
+            val ok = nextFt.meta.rightOwner.parent.contains(t.self)
+            if (ok) Some((ft, t)) else None
+          case _ => None
+        }
+      case _ =>
+        // Optional braces after any token that can start indentation:
+        // =  =>  ?=>  <-  catch  do  else  finally  for
+        // if  match  return  then  throw  try  while  yield
+        val nextFt = nextNonComment(ft)
+        nextFt.meta.rightOwner.parent match {
+          case Some(t: Term.Block) =>
+            val ok = t.tokens.headOption.contains(nextFt.right)
```

Comment:
this is simple enough. it wouldn't match regular `{}` blocks because block is the parent of `{`, but with optional braces, block is the grandparent of the next non-comment token, and that that non-comment token is the first in the block.